"""
üîπ Test Suite for Hash Utilities üîπ

Validates cryptographic hashing, Protobuf encoding, and integrity verification
in the Seigr Hash Utilities module.
"""

import pytest
from unittest.mock import patch, ANY
from src.crypto.hypha_crypt import HyphaCrypt  # ‚úÖ Correctly import HyphaCrypt
from src.crypto.hash_utils import hash_to_protobuf  # ‚úÖ Import from hash_utils
from src.crypto.constants import DEFAULT_HASH_FUNCTION, SUPPORTED_HASH_ALGORITHMS
from src.seigr_protocol.compiled.alerting_pb2 import AlertSeverity
from src.seigr_protocol.compiled.hashing_pb2 import HashAlgorithm
from src.seigr_protocol.compiled.error_handling_pb2 import ErrorResolutionStrategy
from src.logger.secure_logger import secure_logger
from src.crypto.integrity_verification import verify_hash
from src.seigr_protocol.compiled.hashing_pb2 import VerificationStatus  # ‚úÖ Ensure correct import


# üîπ Sample Data for Testing
SAMPLE_DATA = b"Seigr Hash Test Data"
SALT = "test_salt"
HASH_ALGORITHM = "HASH_SEIGR_SENARY"  # ‚úÖ Use correct uppercase format

@pytest.fixture
def hypha_crypt():
    """Fixture to initialize HyphaCrypt instance."""
    return HyphaCrypt(
        data=SAMPLE_DATA,
        segment_id="test_segment",
        hash_depth=3,
        use_senary=True,
    )

### üîπ **Hashing Tests** ###

def test_hypha_hash(hypha_crypt):
    """Test `HyphaCrypt.HASH_SEIGR_SENARY` hashing functionality."""
    hashed_value = hypha_crypt.HASH_SEIGR_SENARY(SAMPLE_DATA, salt=SALT, algorithm=HASH_ALGORITHM)
    assert isinstance(hashed_value, str)  # ‚úÖ Ensure output is a string

def test_hypha_hash_invalid_algorithm(hypha_crypt):
    """Test hashing with an unsupported algorithm should raise ValueError."""
    with pytest.raises(ValueError):
        hypha_crypt.HASH_SEIGR_SENARY(SAMPLE_DATA, salt=SALT, algorithm="UNSUPPORTED_ALGO")

### üîπ **Protobuf Encoding Tests** ###

def test_hash_to_protobuf(hypha_crypt):
    """Test encoding hashed data into a Protobuf format."""
    hash_proto = hash_to_protobuf(SAMPLE_DATA, salt=SALT, algorithm=HASH_ALGORITHM.upper(), version=1)

    assert hash_proto.hash_value, "Hash value should not be empty"
    assert hash_proto.algorithm == HashAlgorithm.HASH_SEIGR_SENARY, "Algorithm should match"
    assert hash_proto.verification_status == VerificationStatus.VERIFICATION_PENDING, "Verification status should be correct"

### üîπ **Logging Tests** ###

@patch.object(secure_logger, "log_audit_event")
def test_hypha_hash_logging(mock_log, hypha_crypt):
    """Test logging on hashing failure."""
    with pytest.raises(ValueError):
        hypha_crypt.HASH_SEIGR_SENARY(SAMPLE_DATA, salt=SALT, algorithm="UNSUPPORTED_ALGO")

    mock_log.assert_any_call(
        severity=AlertSeverity.ALERT_SEVERITY_WARNING,
        category="Hashing",
        message="‚ùå Unsupported hash algorithm detected: UNSUPPORTED_ALGO",
    )

@patch.object(secure_logger, "log_audit_event")
def test_verify_hash_logging(mock_log):
    """Test logging on hash verification failure."""
    verify_hash(SAMPLE_DATA, "invalid_format_hash", salt=SALT)

    mock_log.assert_any_call(
        severity=AlertSeverity.ALERT_SEVERITY_WARNING,  # ‚úÖ Match function's severity
        category="Integrity",  # ‚úÖ Match function's category
        message="SEIGR Hash verification failed.",  # ‚úÖ Match function's message
        log_data={
            "expected_hash": "invalid_format_hash",
            "computed_hash": ANY,  # ‚úÖ Allow any computed hash value
            "match": False,
        },
    )

@patch.object(secure_logger, "log_audit_event")
def test_error_resolution_strategy_logging(mock_log):
    """Test logging includes correct error resolution strategy."""
    verify_hash(SAMPLE_DATA, "invalid_format_hash", salt=SALT)

    mock_log.assert_any_call(
        severity=AlertSeverity.ALERT_SEVERITY_WARNING,  # ‚úÖ Match function's severity (2)
        category="Integrity",  # ‚úÖ Match function's category
        message="SEIGR Hash verification failed.",  # ‚úÖ Match function's message
        log_data={"expected_hash": "invalid_format_hash", "computed_hash": ANY, "match": False},
    )